# SPDX-FileCopyrightText: 2024 Felix Hilgers <contact@fhilgers.com>
#
# SPDX-License-Identifier: Apache-2.0
# SPDX-License-Identifier: MIT

diff --git MODULE.bazel MODULE.bazel
index a7cf644..db9924b 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -10,6 +10,8 @@ bazel_dep(name = "bazel_skylib", version = "1.6.1")
 bazel_dep(name = "rules_java", version = "7.2.0")
 bazel_dep(name = "rules_python", version = "0.23.1")
 bazel_dep(name = "rules_cc", version = "0.0.8")
+bazel_dep(name = "bazel_jar_jar", version = "0.1.5")
+bazel_dep(name = "rules_android", version = "2.0.0")
 
 rules_kotlin_extensions = use_extension("//src/main/starlark/core/repositories:bzlmod_setup.bzl", "rules_kotlin_extensions")
 use_repo(
@@ -19,7 +21,6 @@ use_repo(
     "com_github_jetbrains_kotlin",
     "com_github_pinterest_ktlint",
     "released_rules_kotlin",
-    "rules_android",
 )
 
 # Once the released rules_koltin is defined, configure it.
@@ -68,7 +69,6 @@ maven.install(
         "com.google.dagger:dagger-producers:2.51",
         "javax.annotation:javax.annotation-api:1.3.2",
         "javax.inject:javax.inject:1",
-        "org.pantsbuild:jarjar:1.7.2",
         "org.jetbrains.kotlinx:atomicfu-js:0.15.2",
         "org.jetbrains.kotlinx:kotlinx-serialization-runtime:1.0-M1-1.4.0-rc",
         "dev.zacsweers.autoservice:auto-service-ksp:jar:1.1.0",
diff --git kotlin/internal/jvm/jvm.bzl kotlin/internal/jvm/jvm.bzl
index 05c38af..384bd8d 100644
--- kotlin/internal/jvm/jvm.bzl
+++ kotlin/internal/jvm/jvm.bzl
@@ -490,7 +490,7 @@ _kt_compiler_deps_aspect = aspect(
         "_jarjar": attr.label(
             executable = True,
             cfg = "exec",
-            default = Label("//third_party:jarjar_runner"),
+            default = Label("@bazel_jar_jar//src/main/java/com/github/johnynek/jarjar:app"),
         ),
     },
 )
@@ -572,7 +572,7 @@ Supports the following template values:
         "_jarjar": attr.label(
             executable = True,
             cfg = "exec",
-            default = Label("//third_party:jarjar_runner"),
+            default = Label("@bazel_jar_jar//src/main/java/com/github/johnynek/jarjar:app"),
         ),
     },
     implementation = _kt_compiler_plugin_impl,
diff --git kotlin_rules_maven_install.json kotlin_rules_maven_install.json
index 522919a..6bfd58d 100755
--- kotlin_rules_maven_install.json
+++ kotlin_rules_maven_install.json
@@ -1,7 +1,7 @@
 {
   "__AUTOGENERATED_FILE_DO_NOT_MODIFY_THIS_FILE_MANUALLY": "THERE_IS_NO_DATA_ONLY_ZUUL",
-  "__INPUT_ARTIFACTS_HASH": -1193919207,
-  "__RESOLVED_ARTIFACTS_HASH": 368834518,
+  "__INPUT_ARTIFACTS_HASH": 1684471725,
+  "__RESOLVED_ARTIFACTS_HASH": -564672273,
   "conflict_resolution": {
     "com.google.auto.service:auto-service-annotations:1.0.1": "com.google.auto.service:auto-service-annotations:1.1.1",
     "com.google.guava:guava:27.1-jre": "com.google.guava:guava:33.0.0-jre"
@@ -231,20 +231,6 @@
       },
       "version": "1.3.2"
     },
-    "javax.annotation:jsr250-api": {
-      "shasums": {
-        "jar": "a1a922d0d9b6d183ed3800dfac01d1e1eb159f0e8c6f94736931c1def54a941f",
-        "sources": "025c47d76c60199381be07012a0c5f9e74661aac5bd67f5aec847741c5b7f838"
-      },
-      "version": "1.0"
-    },
-    "javax.enterprise:cdi-api": {
-      "shasums": {
-        "jar": "1f10b2204cc77c919301f20ff90461c3df1b6e6cb148be1c2d22107f4851d423",
-        "sources": "0e7c351dfe05759f84dc3eddaac1da4ef72578b494b53338829d34b12271374f"
-      },
-      "version": "1.0"
-    },
     "javax.inject:javax.inject": {
       "shasums": {
         "jar": "91c77044a50c481636c32d916fd89c9118a72195390452c81065080f957de7ff",
@@ -266,48 +252,6 @@
       },
       "version": "0.2"
     },
-    "org.apache.ant:ant": {
-      "shasums": {
-        "jar": "d81254bcb2e170c9ea16cd418050f3340da1736380a02415c8ddda9a0a0b8a1b",
-        "sources": "369708c550831c8a1a4c1a333e51db30ad69bbcd674f8545b6bce6a4290a9f5e"
-      },
-      "version": "1.9.9"
-    },
-    "org.apache.ant:ant-launcher": {
-      "shasums": {
-        "jar": "02cbe010ceec4acab059acfa48a16dc7b7c430200f2561b6b7c75bcab48f4044",
-        "sources": "ba0f51943f94c3a107f1aff983f180e08e3bc2093643feacfa6d1af025190ae8"
-      },
-      "version": "1.9.9"
-    },
-    "org.apache.commons:commons-lang3": {
-      "shasums": {
-        "jar": "734c8356420cc8e30c795d64fd1fcd5d44ea9d90342a2cc3262c5158fbc6d98b",
-        "sources": "4709f16a9e0f8fd83ae155083d63044e23045aac8f6f0183a2db09f492491b12"
-      },
-      "version": "3.4"
-    },
-    "org.apache.maven:maven-artifact": {
-      "shasums": {
-        "jar": "1f702928f2233c6ecdf308fbd8f2932ea287c7062183d3c8364b0db7e9c4445d",
-        "sources": "8985ed687dc682ff0fef32fb449005cc02eb47f860e36bf8b38b3be3cfa81ef6"
-      },
-      "version": "3.3.9"
-    },
-    "org.apache.maven:maven-model": {
-      "shasums": {
-        "jar": "15abde67fa7ea1e573e1f68c34921e995f0971351aaf1fb96790688ff510efcd",
-        "sources": "9caac5f1fafe59db3b68f3c78dc6f3511720e59b190d3c1de939ce4fefc1bea6"
-      },
-      "version": "3.3.9"
-    },
-    "org.apache.maven:maven-plugin-api": {
-      "shasums": {
-        "jar": "14cae18fd7125901b12fc914e30ea26ad9bd43dbd399dd6e8fcbc6c754ef2c9c",
-        "sources": "5f26fcbf2634be46e9b55a7c95d92e916fc14c55aca96ad29384a0912476bb9e"
-      },
-      "version": "3.3.9"
-    },
     "org.checkerframework:checker-compat-qual": {
       "shasums": {
         "jar": "11d134b245e9cacc474514d2d66b5b8618f8039a1465cdc55bbc0b34e0008b7a",
@@ -322,41 +266,6 @@
       },
       "version": "3.41.0"
     },
-    "org.codehaus.plexus:plexus-classworlds": {
-      "shasums": {
-        "jar": "b2931d41740490a8d931cbe0cfe9ac20deb66cca606e679f52522f7f534c9fd7",
-        "sources": "d087c4c0ff02b035111bb72c72603b2851d126c43da39cc3c73ff45139125bec"
-      },
-      "version": "2.5.2"
-    },
-    "org.codehaus.plexus:plexus-component-annotations": {
-      "shasums": {
-        "jar": "4df7a6a7be64b35bbccf60b5c115697f9ea3421d22674ae67135dde375fcca1f",
-        "sources": "527768d357304e0ad56b74ca77f27ba28b4a456680450ef45a30bfaf613469e6"
-      },
-      "version": "1.5.5"
-    },
-    "org.codehaus.plexus:plexus-utils": {
-      "shasums": {
-        "jar": "0f31c44b275f87e56d46a582ce96d03b9e2ab344cf87c4e268b34d3ad046beab",
-        "sources": "8fb619ac58aaa2a27f9cefce866a9689d41e15a120c2efe0f781b6bcad88caf3"
-      },
-      "version": "3.0.22"
-    },
-    "org.eclipse.sisu:org.eclipse.sisu.inject": {
-      "shasums": {
-        "jar": "66e87705a818da44eb080a7bb1fc431de987754b4f92aa85f69991bfc677d40d",
-        "sources": "739d7228920a97892b9c6b6e4d16799930e8d2439543e71fa7a7c849a39cc8a1"
-      },
-      "version": "0.3.2"
-    },
-    "org.eclipse.sisu:org.eclipse.sisu.plexus": {
-      "shasums": {
-        "jar": "f5cfe0d88e8276971db4ecff0e4186d5f2ec5fdb1b6bb8c2f359fdc4b43eb8b2",
-        "sources": "d8091fc791e3027dd72e51f0dabcd387f27dba8fffa79511edea29f33213a8a8"
-      },
-      "version": "0.3.2"
-    },
     "org.hamcrest:hamcrest-core": {
       "shasums": {
         "jar": "66fdef91e9739348df7a096aa384a5685f4e875584cce89386a7a47251c4d8e9",
@@ -433,34 +342,6 @@
         "sources": "5ceb3d0b41a8eb8b416f28778b27c0e290143b6dc29d887b1cf40e2c0727f096"
       },
       "version": "9.3"
-    },
-    "org.ow2.asm:asm-analysis": {
-      "shasums": {
-        "jar": "e981f8f650c4d900bb033650b18e122fa6b161eadd5f88978d08751f72ee8474",
-        "sources": "57bdf5b407dc122b8f4118e2fbf686719c81f6b7c97598e17ce7a456ea151866"
-      },
-      "version": "7.0"
-    },
-    "org.ow2.asm:asm-commons": {
-      "shasums": {
-        "jar": "fed348ef05958e3e846a3ac074a12af5f7936ef3d21ce44a62c4fa08a771927d",
-        "sources": "7076b9f1cd3fdae003447b9fd546290b9ab76e34f147b2bf0b981ddae86f7053"
-      },
-      "version": "7.0"
-    },
-    "org.ow2.asm:asm-tree": {
-      "shasums": {
-        "jar": "cfd7a0874f9de36a999c127feeadfbfe6e04d4a71ee954d7af3d853f0be48a6c",
-        "sources": "8888c88a9889b1b413d90e440b6b9c894e8d31632a473bd70e683645aba382dd"
-      },
-      "version": "7.0"
-    },
-    "org.pantsbuild:jarjar": {
-      "shasums": {
-        "jar": "0706a455e17b67718abe212e3a77688bbe8260852fc74e3e836d9f2e76d91c27",
-        "sources": "69e0182465b18189294c6a3e3808595ad820b0b5bc2170324b2f6bb37fb1499a"
-      },
-      "version": "1.7.2"
     }
   },
   "dependencies": {
@@ -571,36 +452,9 @@
       "com.squareup:kotlinpoet",
       "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
     ],
-    "javax.enterprise:cdi-api": [
-      "javax.annotation:jsr250-api",
-      "javax.inject:javax.inject"
-    ],
     "junit:junit": [
       "org.hamcrest:hamcrest-core"
     ],
-    "org.apache.ant:ant": [
-      "org.apache.ant:ant-launcher"
-    ],
-    "org.apache.maven:maven-artifact": [
-      "org.apache.commons:commons-lang3",
-      "org.codehaus.plexus:plexus-utils"
-    ],
-    "org.apache.maven:maven-model": [
-      "org.apache.commons:commons-lang3",
-      "org.codehaus.plexus:plexus-utils"
-    ],
-    "org.apache.maven:maven-plugin-api": [
-      "org.apache.maven:maven-artifact",
-      "org.apache.maven:maven-model",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus"
-    ],
-    "org.eclipse.sisu:org.eclipse.sisu.plexus": [
-      "javax.enterprise:cdi-api",
-      "org.codehaus.plexus:plexus-classworlds",
-      "org.codehaus.plexus:plexus-component-annotations",
-      "org.codehaus.plexus:plexus-utils",
-      "org.eclipse.sisu:org.eclipse.sisu.inject"
-    ],
     "org.jetbrains.kotlin:kotlin-reflect": [
       "org.jetbrains.kotlin:kotlin-stdlib"
     ],
@@ -624,23 +478,6 @@
     "org.jetbrains.kotlinx:kotlinx-serialization-runtime": [
       "org.jetbrains.kotlin:kotlin-stdlib",
       "org.jetbrains.kotlin:kotlin-stdlib-common"
-    ],
-    "org.ow2.asm:asm-analysis": [
-      "org.ow2.asm:asm-tree"
-    ],
-    "org.ow2.asm:asm-commons": [
-      "org.ow2.asm:asm",
-      "org.ow2.asm:asm-analysis",
-      "org.ow2.asm:asm-tree"
-    ],
-    "org.ow2.asm:asm-tree": [
-      "org.ow2.asm:asm"
-    ],
-    "org.pantsbuild:jarjar": [
-      "org.apache.ant:ant",
-      "org.apache.maven:maven-plugin-api",
-      "org.ow2.asm:asm",
-      "org.ow2.asm:asm-commons"
     ]
   },
   "skipped": [
@@ -967,19 +804,6 @@
       "javax.annotation.security",
       "javax.annotation.sql"
     ],
-    "javax.annotation:jsr250-api": [
-      "javax.annotation",
-      "javax.annotation.security"
-    ],
-    "javax.enterprise:cdi-api": [
-      "javax.decorator",
-      "javax.enterprise.context",
-      "javax.enterprise.context.spi",
-      "javax.enterprise.event",
-      "javax.enterprise.inject",
-      "javax.enterprise.inject.spi",
-      "javax.enterprise.util"
-    ],
     "javax.inject:javax.inject": [
       "javax.inject"
     ],
@@ -1020,112 +844,6 @@
     "net.ltgt.gradle.incap:incap": [
       "net.ltgt.gradle.incap"
     ],
-    "org.apache.ant:ant": [
-      "org.apache.tools.ant",
-      "org.apache.tools.ant.attribute",
-      "org.apache.tools.ant.dispatch",
-      "org.apache.tools.ant.filters",
-      "org.apache.tools.ant.filters.util",
-      "org.apache.tools.ant.helper",
-      "org.apache.tools.ant.input",
-      "org.apache.tools.ant.listener",
-      "org.apache.tools.ant.loader",
-      "org.apache.tools.ant.property",
-      "org.apache.tools.ant.taskdefs",
-      "org.apache.tools.ant.taskdefs.compilers",
-      "org.apache.tools.ant.taskdefs.condition",
-      "org.apache.tools.ant.taskdefs.cvslib",
-      "org.apache.tools.ant.taskdefs.email",
-      "org.apache.tools.ant.taskdefs.launcher",
-      "org.apache.tools.ant.taskdefs.optional",
-      "org.apache.tools.ant.taskdefs.optional.ccm",
-      "org.apache.tools.ant.taskdefs.optional.clearcase",
-      "org.apache.tools.ant.taskdefs.optional.depend",
-      "org.apache.tools.ant.taskdefs.optional.depend.constantpool",
-      "org.apache.tools.ant.taskdefs.optional.ejb",
-      "org.apache.tools.ant.taskdefs.optional.extension",
-      "org.apache.tools.ant.taskdefs.optional.extension.resolvers",
-      "org.apache.tools.ant.taskdefs.optional.i18n",
-      "org.apache.tools.ant.taskdefs.optional.j2ee",
-      "org.apache.tools.ant.taskdefs.optional.javacc",
-      "org.apache.tools.ant.taskdefs.optional.javah",
-      "org.apache.tools.ant.taskdefs.optional.jlink",
-      "org.apache.tools.ant.taskdefs.optional.jsp",
-      "org.apache.tools.ant.taskdefs.optional.jsp.compilers",
-      "org.apache.tools.ant.taskdefs.optional.native2ascii",
-      "org.apache.tools.ant.taskdefs.optional.net",
-      "org.apache.tools.ant.taskdefs.optional.pvcs",
-      "org.apache.tools.ant.taskdefs.optional.script",
-      "org.apache.tools.ant.taskdefs.optional.sos",
-      "org.apache.tools.ant.taskdefs.optional.testing",
-      "org.apache.tools.ant.taskdefs.optional.unix",
-      "org.apache.tools.ant.taskdefs.optional.vss",
-      "org.apache.tools.ant.taskdefs.optional.windows",
-      "org.apache.tools.ant.taskdefs.rmic",
-      "org.apache.tools.ant.types",
-      "org.apache.tools.ant.types.mappers",
-      "org.apache.tools.ant.types.optional",
-      "org.apache.tools.ant.types.optional.depend",
-      "org.apache.tools.ant.types.resources",
-      "org.apache.tools.ant.types.resources.comparators",
-      "org.apache.tools.ant.types.resources.selectors",
-      "org.apache.tools.ant.types.selectors",
-      "org.apache.tools.ant.types.selectors.modifiedselector",
-      "org.apache.tools.ant.types.spi",
-      "org.apache.tools.ant.util",
-      "org.apache.tools.ant.util.depend",
-      "org.apache.tools.ant.util.facade",
-      "org.apache.tools.ant.util.java15",
-      "org.apache.tools.ant.util.optional",
-      "org.apache.tools.ant.util.regexp",
-      "org.apache.tools.bzip2",
-      "org.apache.tools.mail",
-      "org.apache.tools.tar",
-      "org.apache.tools.zip"
-    ],
-    "org.apache.ant:ant-launcher": [
-      "org.apache.tools.ant.launch"
-    ],
-    "org.apache.commons:commons-lang3": [
-      "org.apache.commons.lang3",
-      "org.apache.commons.lang3.builder",
-      "org.apache.commons.lang3.concurrent",
-      "org.apache.commons.lang3.event",
-      "org.apache.commons.lang3.exception",
-      "org.apache.commons.lang3.math",
-      "org.apache.commons.lang3.mutable",
-      "org.apache.commons.lang3.reflect",
-      "org.apache.commons.lang3.text",
-      "org.apache.commons.lang3.text.translate",
-      "org.apache.commons.lang3.time",
-      "org.apache.commons.lang3.tuple"
-    ],
-    "org.apache.maven:maven-artifact": [
-      "org.apache.maven.artifact",
-      "org.apache.maven.artifact.handler",
-      "org.apache.maven.artifact.metadata",
-      "org.apache.maven.artifact.repository",
-      "org.apache.maven.artifact.repository.layout",
-      "org.apache.maven.artifact.repository.metadata",
-      "org.apache.maven.artifact.resolver",
-      "org.apache.maven.artifact.resolver.filter",
-      "org.apache.maven.artifact.versioning",
-      "org.apache.maven.repository",
-      "org.apache.maven.repository.legacy.metadata"
-    ],
-    "org.apache.maven:maven-model": [
-      "org.apache.maven.model",
-      "org.apache.maven.model.io.xpp3",
-      "org.apache.maven.model.merge"
-    ],
-    "org.apache.maven:maven-plugin-api": [
-      "org.apache.maven.monitor.logging",
-      "org.apache.maven.plugin",
-      "org.apache.maven.plugin.descriptor",
-      "org.apache.maven.plugin.lifecycle",
-      "org.apache.maven.plugin.lifecycle.io.xpp3",
-      "org.apache.maven.plugin.logging"
-    ],
     "org.checkerframework:checker-compat-qual": [
       "org.checkerframework.checker.nullness.compatqual"
     ],
@@ -1161,60 +879,6 @@
       "org.checkerframework.dataflow.qual",
       "org.checkerframework.framework.qual"
     ],
-    "org.codehaus.plexus:plexus-classworlds": [
-      "org.codehaus.classworlds",
-      "org.codehaus.plexus.classworlds",
-      "org.codehaus.plexus.classworlds.launcher",
-      "org.codehaus.plexus.classworlds.realm",
-      "org.codehaus.plexus.classworlds.strategy"
-    ],
-    "org.codehaus.plexus:plexus-component-annotations": [
-      "org.codehaus.plexus.component.annotations"
-    ],
-    "org.codehaus.plexus:plexus-utils": [
-      "org.codehaus.plexus.util",
-      "org.codehaus.plexus.util.cli",
-      "org.codehaus.plexus.util.cli.shell",
-      "org.codehaus.plexus.util.dag",
-      "org.codehaus.plexus.util.introspection",
-      "org.codehaus.plexus.util.io",
-      "org.codehaus.plexus.util.reflection",
-      "org.codehaus.plexus.util.xml",
-      "org.codehaus.plexus.util.xml.pull"
-    ],
-    "org.eclipse.sisu:org.eclipse.sisu.inject": [
-      "org.eclipse.sisu",
-      "org.eclipse.sisu.bean",
-      "org.eclipse.sisu.inject",
-      "org.eclipse.sisu.launch",
-      "org.eclipse.sisu.osgi",
-      "org.eclipse.sisu.space",
-      "org.eclipse.sisu.space.asm",
-      "org.eclipse.sisu.wire",
-      "org.sonatype.inject"
-    ],
-    "org.eclipse.sisu:org.eclipse.sisu.plexus": [
-      "org.codehaus.plexus",
-      "org.codehaus.plexus.component",
-      "org.codehaus.plexus.component.composition",
-      "org.codehaus.plexus.component.configurator",
-      "org.codehaus.plexus.component.configurator.converters",
-      "org.codehaus.plexus.component.configurator.converters.basic",
-      "org.codehaus.plexus.component.configurator.converters.composite",
-      "org.codehaus.plexus.component.configurator.converters.lookup",
-      "org.codehaus.plexus.component.configurator.converters.special",
-      "org.codehaus.plexus.component.configurator.expression",
-      "org.codehaus.plexus.component.factory",
-      "org.codehaus.plexus.component.repository",
-      "org.codehaus.plexus.component.repository.exception",
-      "org.codehaus.plexus.configuration",
-      "org.codehaus.plexus.configuration.xml",
-      "org.codehaus.plexus.context",
-      "org.codehaus.plexus.logging",
-      "org.codehaus.plexus.logging.console",
-      "org.codehaus.plexus.personality.plexus.lifecycle.phase",
-      "org.eclipse.sisu.plexus"
-    ],
     "org.hamcrest:hamcrest-core": [
       "org.hamcrest",
       "org.hamcrest.core",
@@ -1346,20 +1010,6 @@
     "org.ow2.asm:asm": [
       "org.objectweb.asm",
       "org.objectweb.asm.signature"
-    ],
-    "org.ow2.asm:asm-analysis": [
-      "org.objectweb.asm.tree.analysis"
-    ],
-    "org.ow2.asm:asm-commons": [
-      "org.objectweb.asm.commons"
-    ],
-    "org.ow2.asm:asm-tree": [
-      "org.objectweb.asm.tree"
-    ],
-    "org.pantsbuild:jarjar": [
-      "org.pantsbuild.jarjar",
-      "org.pantsbuild.jarjar.misplaced",
-      "org.pantsbuild.jarjar.util"
     ]
   },
   "repositories": {
@@ -1427,42 +1077,16 @@
       "dev.zacsweers.autoservice:auto-service-ksp:jar:sources",
       "javax.annotation:javax.annotation-api",
       "javax.annotation:javax.annotation-api:jar:sources",
-      "javax.annotation:jsr250-api",
-      "javax.annotation:jsr250-api:jar:sources",
-      "javax.enterprise:cdi-api",
-      "javax.enterprise:cdi-api:jar:sources",
       "javax.inject:javax.inject",
       "javax.inject:javax.inject:jar:sources",
       "junit:junit",
       "junit:junit:jar:sources",
       "net.ltgt.gradle.incap:incap",
       "net.ltgt.gradle.incap:incap:jar:sources",
-      "org.apache.ant:ant",
-      "org.apache.ant:ant-launcher",
-      "org.apache.ant:ant-launcher:jar:sources",
-      "org.apache.ant:ant:jar:sources",
-      "org.apache.commons:commons-lang3",
-      "org.apache.commons:commons-lang3:jar:sources",
-      "org.apache.maven:maven-artifact",
-      "org.apache.maven:maven-artifact:jar:sources",
-      "org.apache.maven:maven-model",
-      "org.apache.maven:maven-model:jar:sources",
-      "org.apache.maven:maven-plugin-api",
-      "org.apache.maven:maven-plugin-api:jar:sources",
       "org.checkerframework:checker-compat-qual",
       "org.checkerframework:checker-compat-qual:jar:sources",
       "org.checkerframework:checker-qual",
       "org.checkerframework:checker-qual:jar:sources",
-      "org.codehaus.plexus:plexus-classworlds",
-      "org.codehaus.plexus:plexus-classworlds:jar:sources",
-      "org.codehaus.plexus:plexus-component-annotations",
-      "org.codehaus.plexus:plexus-component-annotations:jar:sources",
-      "org.codehaus.plexus:plexus-utils",
-      "org.codehaus.plexus:plexus-utils:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.inject",
-      "org.eclipse.sisu:org.eclipse.sisu.inject:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus:jar:sources",
       "org.hamcrest:hamcrest-core",
       "org.hamcrest:hamcrest-core:jar:sources",
       "org.jetbrains.kotlin:kotlin-reflect",
@@ -1484,15 +1108,7 @@
       "org.jetbrains:annotations",
       "org.jetbrains:annotations:jar:sources",
       "org.ow2.asm:asm",
-      "org.ow2.asm:asm-analysis",
-      "org.ow2.asm:asm-analysis:jar:sources",
-      "org.ow2.asm:asm-commons",
-      "org.ow2.asm:asm-commons:jar:sources",
-      "org.ow2.asm:asm-tree",
-      "org.ow2.asm:asm-tree:jar:sources",
-      "org.ow2.asm:asm:jar:sources",
-      "org.pantsbuild:jarjar",
-      "org.pantsbuild:jarjar:jar:sources"
+      "org.ow2.asm:asm:jar:sources"
     ],
     "https://maven.google.com/": [
       "com.google.auto.service:auto-service",
@@ -1558,42 +1174,16 @@
       "dev.zacsweers.autoservice:auto-service-ksp:jar:sources",
       "javax.annotation:javax.annotation-api",
       "javax.annotation:javax.annotation-api:jar:sources",
-      "javax.annotation:jsr250-api",
-      "javax.annotation:jsr250-api:jar:sources",
-      "javax.enterprise:cdi-api",
-      "javax.enterprise:cdi-api:jar:sources",
       "javax.inject:javax.inject",
       "javax.inject:javax.inject:jar:sources",
       "junit:junit",
       "junit:junit:jar:sources",
       "net.ltgt.gradle.incap:incap",
       "net.ltgt.gradle.incap:incap:jar:sources",
-      "org.apache.ant:ant",
-      "org.apache.ant:ant-launcher",
-      "org.apache.ant:ant-launcher:jar:sources",
-      "org.apache.ant:ant:jar:sources",
-      "org.apache.commons:commons-lang3",
-      "org.apache.commons:commons-lang3:jar:sources",
-      "org.apache.maven:maven-artifact",
-      "org.apache.maven:maven-artifact:jar:sources",
-      "org.apache.maven:maven-model",
-      "org.apache.maven:maven-model:jar:sources",
-      "org.apache.maven:maven-plugin-api",
-      "org.apache.maven:maven-plugin-api:jar:sources",
       "org.checkerframework:checker-compat-qual",
       "org.checkerframework:checker-compat-qual:jar:sources",
       "org.checkerframework:checker-qual",
       "org.checkerframework:checker-qual:jar:sources",
-      "org.codehaus.plexus:plexus-classworlds",
-      "org.codehaus.plexus:plexus-classworlds:jar:sources",
-      "org.codehaus.plexus:plexus-component-annotations",
-      "org.codehaus.plexus:plexus-component-annotations:jar:sources",
-      "org.codehaus.plexus:plexus-utils",
-      "org.codehaus.plexus:plexus-utils:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.inject",
-      "org.eclipse.sisu:org.eclipse.sisu.inject:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus:jar:sources",
       "org.hamcrest:hamcrest-core",
       "org.hamcrest:hamcrest-core:jar:sources",
       "org.jetbrains.kotlin:kotlin-reflect",
@@ -1615,15 +1205,7 @@
       "org.jetbrains:annotations",
       "org.jetbrains:annotations:jar:sources",
       "org.ow2.asm:asm",
-      "org.ow2.asm:asm-analysis",
-      "org.ow2.asm:asm-analysis:jar:sources",
-      "org.ow2.asm:asm-commons",
-      "org.ow2.asm:asm-commons:jar:sources",
-      "org.ow2.asm:asm-tree",
-      "org.ow2.asm:asm-tree:jar:sources",
-      "org.ow2.asm:asm:jar:sources",
-      "org.pantsbuild:jarjar",
-      "org.pantsbuild:jarjar:jar:sources"
+      "org.ow2.asm:asm:jar:sources"
     ],
     "https://repo1.maven.org/maven2/": [
       "com.google.auto.service:auto-service",
@@ -1689,42 +1271,16 @@
       "dev.zacsweers.autoservice:auto-service-ksp:jar:sources",
       "javax.annotation:javax.annotation-api",
       "javax.annotation:javax.annotation-api:jar:sources",
-      "javax.annotation:jsr250-api",
-      "javax.annotation:jsr250-api:jar:sources",
-      "javax.enterprise:cdi-api",
-      "javax.enterprise:cdi-api:jar:sources",
       "javax.inject:javax.inject",
       "javax.inject:javax.inject:jar:sources",
       "junit:junit",
       "junit:junit:jar:sources",
       "net.ltgt.gradle.incap:incap",
       "net.ltgt.gradle.incap:incap:jar:sources",
-      "org.apache.ant:ant",
-      "org.apache.ant:ant-launcher",
-      "org.apache.ant:ant-launcher:jar:sources",
-      "org.apache.ant:ant:jar:sources",
-      "org.apache.commons:commons-lang3",
-      "org.apache.commons:commons-lang3:jar:sources",
-      "org.apache.maven:maven-artifact",
-      "org.apache.maven:maven-artifact:jar:sources",
-      "org.apache.maven:maven-model",
-      "org.apache.maven:maven-model:jar:sources",
-      "org.apache.maven:maven-plugin-api",
-      "org.apache.maven:maven-plugin-api:jar:sources",
       "org.checkerframework:checker-compat-qual",
       "org.checkerframework:checker-compat-qual:jar:sources",
       "org.checkerframework:checker-qual",
       "org.checkerframework:checker-qual:jar:sources",
-      "org.codehaus.plexus:plexus-classworlds",
-      "org.codehaus.plexus:plexus-classworlds:jar:sources",
-      "org.codehaus.plexus:plexus-component-annotations",
-      "org.codehaus.plexus:plexus-component-annotations:jar:sources",
-      "org.codehaus.plexus:plexus-utils",
-      "org.codehaus.plexus:plexus-utils:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.inject",
-      "org.eclipse.sisu:org.eclipse.sisu.inject:jar:sources",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus",
-      "org.eclipse.sisu:org.eclipse.sisu.plexus:jar:sources",
       "org.hamcrest:hamcrest-core",
       "org.hamcrest:hamcrest-core:jar:sources",
       "org.jetbrains.kotlin:kotlin-reflect",
@@ -1746,15 +1302,7 @@
       "org.jetbrains:annotations",
       "org.jetbrains:annotations:jar:sources",
       "org.ow2.asm:asm",
-      "org.ow2.asm:asm-analysis",
-      "org.ow2.asm:asm-analysis:jar:sources",
-      "org.ow2.asm:asm-commons",
-      "org.ow2.asm:asm-commons:jar:sources",
-      "org.ow2.asm:asm-tree",
-      "org.ow2.asm:asm-tree:jar:sources",
-      "org.ow2.asm:asm:jar:sources",
-      "org.pantsbuild:jarjar",
-      "org.pantsbuild:jarjar:jar:sources"
+      "org.ow2.asm:asm:jar:sources"
     ]
   },
   "version": "2"
diff --git third_party/BUILD third_party/BUILD
index 624fe5f..fcce02c 100644
--- third_party/BUILD
+++ third_party/BUILD
@@ -66,17 +66,10 @@ java_library(
     ],
 )
 
-java_binary(
-    name = "jarjar_runner",
-    main_class = "org.pantsbuild.jarjar.Main",
-    visibility = ["//visibility:public"],
-    runtime_deps = ["@kotlin_rules_maven//:org_pantsbuild_jarjar"],
-)
-
 # TODO(bazelbuild/rules_kotlin/issues/273): Remove android_sdk import.
 java_import(
     name = "android_sdk",
-    jars = ["@bazel_tools//tools/android:android_jar"],
+    jars = ["@rules_android//tools/android:android_jar"],
     neverlink = 1,
     visibility = ["//visibility:public"],
 )
@@ -97,7 +90,7 @@ release_archive(
         "jarjar.bzl",
     ],
     src_map = {
-        "jarjar_runner_deploy.jar": "jarjar.jar",
+        "@bazel_jar_jar//src/main/java/com/github/johnynek/jarjar:app_deploy.jar": "jarjar.jar",
         "BUILD.release.bazel": "BUILD.bazel",
     },
 )
diff --git third_party/jarjar.bzl third_party/jarjar.bzl
index 0eedc25..6b122b9 100644
--- third_party/jarjar.bzl
+++ third_party/jarjar.bzl
@@ -50,7 +50,7 @@ jar_jar = rule(
         "jarjar_runner": attr.label(
             executable = True,
             cfg = "exec",
-            default = Label("//third_party:jarjar_runner"),
+            default = Label("@bazel_jar_jar//src/main/java/com/github/johnynek/jarjar:app"),
         ),
     },
     outputs = {
